platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/consensus/AncestorSearch.java

```diff
@@ -0,0 +1,92 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.consensus;
+
+import com.swirlds.platform.internal.EventImpl;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Predicate;
+
+/** Used for searching for ancestors events in a hashgraph */
+public class AncestorSearch {
+    /** the iterator that returns ancestors */
+    private final AncestorIterator iterator;
+
+    public AncestorSearch() {
+        this(new EventVisitedMark());
+    }
+
+    /**
+     * Create a new ancestor search with the given mark
+     * @param mark the instance to use to mark events as visited
+     */
+    public AncestorSearch(@NonNull final EventVisitedMark mark) {
+        this.iterator = new AncestorIterator(mark);
+    }
+
+    /**
+     * The set of ancestors of the given event (the root of the search) for which valid is true.
+     * This will not include a valid ancestor that is only reachable through invalid ancestors.
+     *
+     * @param root the root event whose ancestors should be searched
+     * @param valid do a depth-first search, but backtrack from any event e where valid(e)==false
+     */
+    public @NonNull AncestorIterator search(@NonNull final EventImpl root, @NonNull final Predicate<EventImpl> valid) {
+        iterator.search(root, valid);
+        return iterator;
+    }
+
+    /**
+     * Finds events that ancestors to all the supplied events. It will only traverse valid events,
+     * where valid is defined by the predicate supplied. This method will also populate {@link
+     * EventImpl#getRecTimes()} for the events returned.
+     *
+     * @param events the events whose ancestors we are searching for
+     * @param valid checks if the event should be part of the search
+     * @return a list of all common ancestors
+     */
+    public @NonNull List<EventImpl> commonAncestorsOf(
+            @NonNull final List<EventImpl> events, @NonNull final Predicate<EventImpl> valid) {
+        // each event visited by iterator from at least one of the provided events
+        final ArrayList<EventImpl> visited = new ArrayList<>();
+        // Do a non-recursive search of the hashgraph, without using the Java stack, and being
+        // efficient when it's a DAG that isn't a tree.
+        for (final EventImpl e : events) {
+            final AncestorIterator validAncestors = search(e, valid);
+            while (validAncestors.hasNext()) {
+                final EventImpl event = validAncestors.next();
+                if (event.getRecTimes() == null) {
+                    event.setRecTimes(new ArrayList<>());
+                    visited.add(event);
+                }
+                event.getRecTimes().add(validAncestors.getTime());
```

Comment:
I'm still confused about the description of these times. This is very important, as these times determine the consensus time of each event. Perhaps let's chat about this and try to come up with a clear way to describe what is happening.

Reply:
sure. again, this is unchanged code, just repackaged